examples/step-9/step-9.cc

```diff
@@ -800,58 +712,51 @@ namespace Step9
                        scratch_data.fe_face_values.shape_value(i, q_point) *
                        scratch_data.fe_face_values.shape_value(j, q_point) *
                        scratch_data.fe_face_values.JxW(q_point));
 
                   copy_data.cell_rhs(i) -=
-                    (face_advection_directions[q_point] *
+                    (scratch_data.face_advection_directions[q_point] *
                      scratch_data.fe_face_values.normal_vector(q_point) *
-                     face_boundary_values[q_point] *
+                     scratch_data.face_boundary_values[q_point] *
                      scratch_data.fe_face_values.shape_value(i, q_point) *
                      scratch_data.fe_face_values.JxW(q_point));
                 }
         }
 
-
-    // Now go on by transferring the local contributions to the system of
-    // equations into the global objects. The first step was to obtain the
-    // global indices of the degrees of freedom on this cell.
+    // The final piece of information the copy routine needs is the cell DoF
+    // indices, so we end by writing them to the local array:
     cell->get_dof_indices(copy_data.local_dof_indices);
   }
 
 
 
   // The second function we needed to write was the one that copies
-  // the local contributions the previous function has computed and
-  // put into the copy data object, into the global matrix and right
+  // the local contributions the previous function computed (and
+  // put into the AssemblyCopyData object) into the global matrix and right
   // hand side vector objects. This is essentially what we always had
   // as the last block of code when assembling something on every
   // cell. The following should therefore be pretty obvious:
   template <int dim>
   void
   AdvectionProblem<dim>::copy_local_to_global(const AssemblyCopyData &copy_data)
   {
-    for (unsigned int i = 0; i < copy_data.local_dof_indices.size(); ++i)
-      {
-        for (unsigned int j = 0; j < copy_data.local_dof_indices.size(); ++j)
-          system_matrix.add(copy_data.local_dof_indices[i],
-                            copy_data.local_dof_indices[j],
-                            copy_data.cell_matrix(i, j));
-
-        system_rhs(copy_data.local_dof_indices[i]) += copy_data.cell_rhs(i);
-      }
+    hanging_node_constraints.distribute_local_to_global(
+      copy_data.cell_matrix,
+      copy_data.cell_rhs,
+      copy_data.local_dof_indices,
+      system_matrix,
+      system_rhs);
   }
 
-
-
-  // Following is the function that solves the linear system of equations. As
-  // the system is no more symmetric positive definite as in all the previous
-  // examples, we can't use the Conjugate Gradients method anymore. Rather, we
-  // use a solver that is tailored to nonsymmetric systems like the one at
-  // hand, the BiCGStab method. As preconditioner, we use the Jacobi method.
+  // Here comes the linear solver routine. As the system is no longer
```

Comment:
Didn't we have a conversation at one point in which you said that "Here comes..." is a Germanism? ;-)

Reply:
It is :)

I hope you don't mind that I added it here: that phrase (or "next comes") was used a lot in this tutorial. I added it here (instead of the original, more complex sentence) to keep the style consistent.

Reply:
Oh, I don't mind at all :-)